src/EventListener/OrderListener.php

```diff
@@ -55,4 +67,56 @@ public function redirectToHome(Order\Event\UpdateFailedEvent $event)
 
 		$this->get('event.dispatcher')->dispatch(Page\Event\Event::RENDER_SET_RESPONSE, $redirectEvent);
 	}
+
+	/**
+	 * @param Order\Event\CancelEvent $event
+	 */
+	public function setOrderRefundController(Order\Event\CancelEvent $event)
+	{
+		$this->_setRefundController($event, 'order');
+	}
+
+	/**
+	 * @param Order\Event\CancelEvent $event
+	 */
+	public function setItemRefundController(Order\Event\CancelEvent $event)
+	{
+		$this->_setRefundController($event, 'item');
+	}
+
+	/**
+	 * Set controller and parameters to process refund via gateway. Can configure cancellation for
+	 * orders or individual items by giving 'order' or 'item' as the second parameter
+	 *
+	 * @param Order\Event\CancelEvent $event
+	 * @param $type
+	 */
+	private function _setRefundController(Order\Event\CancelEvent $event, $type)
+	{
+		$types = ['order', 'item'];
+		if (!in_array($type, $types)) {
+			throw new \LogicException('Invalid refund type, must be in array: ' . implode(', ', $types));
+		}
+
+		$gateway = $this->get('gateway');
```

Comment:
The gateway service is deprecated in this very PR. Can we default to an error or something if no gateway is found? This would be exceptional behaviour.